crates/ra_hir/src/ty/infer.rs

```diff
@@ -63,6 +63,24 @@ enum ExprOrPatId {
 
 impl_froms!(ExprOrPatId: ExprId, PatId);
 
+/// Binding modes inferred for patterns.
+/// https://doc.rust-lang.org/reference/patterns.html#binding-modes
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum BindingMode {
```

Comment:
I wasn't sure where to put this new `enum`. As it is only used by inference, I left it here for now.